cumulus/client/consensus/aura/src/collators/slot_based/block_builder_task.rs

```diff
@@ -0,0 +1,453 @@
+			continue;
+		};
+
+		let max_pov_size = match relay_client
+			.persisted_validation_data(relay_parent, para_id, OccupiedCoreAssumption::Included)
+			.await
+		{
+			Ok(None) => continue,
+			Ok(Some(pvd)) => pvd.max_pov_size,
+			Err(err) => {
+				tracing::error!(target: crate::LOG_TARGET, ?err, "Failed to gather information from relay-client");
+				continue
+			},
+		};
+
+		let (included_block, parent) =
+			match find_parent(relay_parent, para_id, &*para_backend, &relay_client).await {
+				Some(value) => value,
+				None => continue,
+			};
+
+		let parent_header = parent.header;
+		let parent_hash = parent.hash;
+
+		// We mainly call this to inform users at genesis if there is a mismatch with the
+		// on-chain data.
+		collator.collator_service().check_block_status(parent_hash, &parent_header);
+
+		let slot_claim = match can_build_upon::<_, _, P>(
+			para_slot.slot,
+			para_slot.timestamp,
+			parent_hash,
+			included_block,
+			&*para_client,
+			&keystore,
+		)
+		.await
+		{
+			Some(slot) => slot,
+			None => continue,
+		};
+
+		tracing::debug!(
+			target: crate::LOG_TARGET,
+			?core_index,
+			slot = ?para_slot.slot,
+			unincluded_segment_len = parent.depth,
+			relay_parent = %relay_parent,
+			included = %included_block,
+			parent = %parent_hash,
+			"Building block."
+		);
+
+		let validation_data = PersistedValidationData {
+			parent_head: parent_header.encode().into(),
+			relay_parent_number: *relay_parent_header.number(),
+			relay_parent_storage_root: *relay_parent_header.state_root(),
+			max_pov_size,
+		};
+
+		let (parachain_inherent_data, other_inherent_data) = match collator
+			.create_inherent_data(
+				relay_parent,
+				&validation_data,
+				parent_hash,
+				slot_claim.timestamp(),
+			)
+			.await
+		{
+			Err(err) => {
+				tracing::error!(target: crate::LOG_TARGET, ?err);
+				break
+			},
+			Ok(x) => x,
+		};
+
+		let validation_code_hash = match code_hash_provider.code_hash_at(parent_hash) {
+			None => {
+				tracing::error!(target: crate::LOG_TARGET, ?parent_hash, "Could not fetch validation code hash");
+				break
+			},
+			Some(v) => v,
+		};
+
+		check_validation_code_or_log(&validation_code_hash, para_id, &relay_client, relay_parent)
+			.await;
+
+		let Ok(Some(candidate)) = collator
+			.build_block_and_import(
+				&parent_header,
+				&slot_claim,
+				None,
+				(parachain_inherent_data, other_inherent_data),
+				authoring_duration,
+				// Set the block limit to 50% of the maximum PoV size.
+				//
+				// TODO: If we got benchmarking that includes the proof size,
+				// we should be able to use the maximum pov size.
+				(validation_data.max_pov_size / 2) as usize,
+			)
+			.await
+		else {
+			tracing::error!(target: crate::LOG_TARGET, "Unable to build block at slot.");
+			continue;
+		};
+
+		let new_block_hash = candidate.block.header().hash();
+
+		// Announce the newly built block to our peers.
+		collator.collator_service().announce_block(new_block_hash, None);
```

Comment:
Is correct to announce the block here?
I was thinking about announcing it in the collator task if `build_collation` goes well.
(I see that the "lookahead" does it in this way (i.e. announces if collation goes well)

Reply:
Hmm indeed it is a change to announce it here. Its a bit of a hassle. If we are unable to build the collation the block can not be submitted to the relay chain and has no chance of getting included. On the other hand, the faster we announce the faster the next block builder can start importing and building. In theory, the collation building can fail but I would say its quite unlikely, we are just fetching some data and compressing the PoV.
Will think about it.

Reply:
Yeah it is fine to announce it here directly. The set is restricted and thus, others can reject if there are "too many" blocks. The runtime also ensures that we are not building too far ahead into the future.